### ISSUE TITLE
Exit Node (without accept-dns) breaks DNS Config from web ui

### ISSUE THREAD

**Issue by Zocker1999NET** [tailscale/tailscale#5875]
### What is the issue?

I discovered that Tailscale does not resolve my DNS requests as expected (through my pi-hole which is set as the only Global Nameserver) if I use an exit node where "accept-dns" is disabled (for other reasons).

I did not expected that setting "--accept-dns=false" on an exit node does modify the behavior of clients using this exit node but with "--accept-dns=true". I understand & accept the concept of using an exit node should let clients feel like they are in the same network as the given exit node. However, I think this should primarily apply to users which do not configure a global DNS server or did not enable "Override local DNS" in the web UI, as I did both (see at steps to reproduce).

For now to have both an exit node & DNS as I expect it, I either need to
- set "--accept-dns=true" on the server, which I cannot do, because the server's applications should not rely on using my pi-hole at home.
- use another exit node without, which heavily slows down my network connection.
- disable the exit node, which is a major privacy issue for me in cellular networks or public networks.
- ignore my DNS setup, which is a privacy & security issue as pi-hole is set up to block domains containing ads, malware and so on.
As so, I first though tailscale is perfect as it allowed me to have secure DNS requests but a fast network speed because I don't route my network connections through my Pi at home but through an hosted server. However, this limitation makes tailscale nearly unusable to me.

### Steps to reproduce

My setup looks as follows:
- I have a pi-hole available (at 10.0.0.10) at home through a subnet router (called `pi`).
  - This should not become my exit node as it has a very slow network speed, but my home network is fast enough to route DNS traffic through it.
  - I want to use my pi-hole as a DNS blocker.
- I have an exit node server (called `server`), which is hosted by a provider.
  - This server should be my exit node as it has a really fast network speed.
  - This server is used for multiple purposes. It should not depend on the pi-hole DNS as this would break many services. That's why "accept-dns" is disabled for it.
- Tailscale DNS Config (web ui): Global Nameserver is set to only 10.0.0.10 (my pi-hole). I enabled "Override local DNS".

Given that `ads.com` is blocked by the pi-hole, if I configure tailscale on my client with said exit node enabled, dns requests are not handled as I would not assume it:
```
$ tailscale up --reset --accept-dns=true --exit-node=server --exit-node-allow-lan-access=true
$ dig ads.com +short
199.59.243.60
```

However, if I disable the exit node (or choose another one without "--accept-dns=false"), DNS requests get handled as I expect (pi-hole answers with `0.0.0.0` to indicate a blocked answer):
```
$ tailscale up --reset --accept-dns=true
$ dig ads.com +short
0.0.0.0
```

### Are there any recent changes that introduced the issue?

Don't know about any.

### OS

Linux

### OS version

Debian GNU/Linux testing/unstable

### Tailscale version

1.30.2

### Bug report

BUG-61ed798e7e05dd0ab5418968cc9693edc8b54953071b2f018aac499dabadd086-20221008113419Z-2be646c65961d209

**Comment by Zocker1999NET** [comment]
A solution which should be simple to implement (but should be configure on all clients) might be: "--exit-node-use-dns=(true|false)" which allows users to choose from following behaviors:
- if set to true, use DNS as the exit node (the default, the current behavior)
- it set to false, use DNS as yourself would do, de facto ignoring the exit node's DNS config (which I expected from the start)

I will look into if this is possible to implement. If so, it would be good to know if such change would be accepted into this repository or not.

**Comment by DentonGentry** [comment]
A client device using an exit node is intended to "teleport to the location of the exit node" as fully as possible. All egress should be from the exit node and all DNS handling should happen at the location of the exit node (for any Geo IP and/or Anycast behaviors).

--------

>A solution which should be simple to implement (but should be configure on all clients) might be: "--exit-node-use-dns=(true|false)"

This proposes a larger change than might be obvious. We allow a DNS configuration in https://login.tailscale.com/admin/dns with "Override local DNS" turned off, meaning there are no configured global DNS servers.

On some platforms once we enable an exit node, we can no longer access the platform's configured DNS servers from inside the sandbox in which tailscaled runs. We cannot implement the behavior of having "Override local DNS" turned off. Forwarding all queries to the exit node both meets the goal of teleporting to the exit node's location as well as makes it possible to function in these platforms.

**Comment by Zocker1999NET** [comment]
> This proposes a larger change than might be obvious.

Can you please look at my PR (#5876)? I implemented said flag quite fast and it works for me good enough on Linux. It still requires testing on other OSes & adding test cases for the new flag & fixing a bug on upgrading (see PR). Are there any edge cases I do not know about by implementing it this way?

> On some platforms once we enable an exit node, we can no longer access the platform's configured DNS servers from inside the sandbox in which tailscaled runs. We cannot implement the behavior of having "Override local DNS" turned off. Forwarding all queries to the exit node both meets the goal of teleporting to the exit node's location as well as makes it possible to function in these platforms.

As I understand it: The flag as I implemented it should not worsen nor improve these problems on such platforms. If using an exit node but with "--exit-node-use-dns=false", Tailscale should continue to fallback to the "global nameserver" setting if it does not know the local DNS resolver (same as before & independent of "Override local DNS" being on or off). If that's not possible, because no "global nameserver" is given, then yes, Tailscale would not have any DNS server to ask from. However then, my new setting may be unavailable on such platforms as long as no "global nameserver" is set. Can you name the affected platforms?

**Comment by DentonGentry** [comment]
>Can you please look at my PR (https://github.com/tailscale/tailscale/pull/5876)?

We're not going to handle this as a code review. You're welcome to implement whatever behavior you wish in your fork.

We have declined to provide an ever-increasing series of knobs for every possible circumstance. We can neither test, document, nor support this, especially when the individual flags which solve each individual problem begin interacting.

When using an exit node, the desired behavior of the feature is to teleport to the location of the exit node. DNS queries are forwarded to the exit node in order to get the GeoIP behavior of initiating DNS queries from the location of the exit node.

> Can you name the affected platforms?

macOS and iOS, when an exit node is activated.

**Comment by DentonGentry** [comment]
Can I ask why the exit node has `--accept-dns=false`? I'd rather solve that problem automatically, without a manual flag.

One possible relevant item: prior to 1.30, nodes in AWS often turned off DNS because certain AWS features like NFS mounting EBS only worked if one used the AWS-supplied DNS servers.

In 1.30 if tailscaled can determine that it is running in AWS, it automatically adds the AWS DNS server for compute.internal.

**Comment by Zocker1999NET** [comment]
> Can I ask why the exit node has --accept-dns=false?
I want my clients to enforce them to only use my Pi-Hole, which I host at home. If that goes down (e.g. because of bad ISP), I can fix my clients manually or wait until Pi-Hole is up again. That's why I configured "Override local DNS". However my server hosts multiple services (Nextcloud, Gitea, Matrix upcoming, Mail upcoming …), which are not only used by me, and often require DNS, and I don't want my server (which is hosted by [Hetzner](https://www.hetzner.com/) to rely on the connectivity of my Pi-Hole nor being slowed down.

**Comment by Zocker1999NET** [comment]
> When using an exit node, the desired behavior of the feature is to teleport to the location of the exit node. DNS queries are forwarded to the exit node in order to get the GeoIP behavior of initiating DNS queries from the location of the exit node.

But IMO this does not require handling them in the extra way by handing them over to the exit node's daemon using DNS-over-HTTP, like it is implemented now. It should be sufficient to just send them through the tunnel to the exit like any other packet, because it still has the source IP of the Exit Node (except when SNAT is disabled, but that already stands for an heavily custom-configured network). So any resolver will still think, that the exit node has issued the DNS request. No custom DNS-over-HTTP or involvement of the other Tailscale daemon (aside from forwarding packages) required. Do you know, why this is implemented?

**Comment by DentonGentry** [comment]
> It should be sufficient to just send them through the tunnel to the exit like any other packet

A couple reasons:

1\
Assume the exit node is in Europe and gets a DNS server via DHCP from an ISP in Europe. The device using the exit node is in Canada and gets a DNS server via DHCP from the ISP in Canada.

Sending to the IP address of the Canadian DNS server by way of the exit node in Europe isn't the right behavior. The intent is that the device teleport to the location of the exit node and use the DNS servers as they would be at that location.

2\
The macOS sandbox issue. If you are on a macOS device with an exit node enabled and "Override local DNS" turned OFF, so that you need to read out the DNS server from the platform, in theory you could find out the DNS server's IP address and then send the packet over to the exit node.

However the sandbox we run in makes it impossible to read out the DNS server address. We can't find out what the address to send to is.

**Comment by Zocker1999NET** [comment]
> Sending to the IP address of the Canadian DNS server by way of the exit node in Europe isn't the right behavior.

Aaahhh, okay, I didn't thought about that case in the first place, sorry. It's kind of normal for me to ignore my ISPs DNS servers because they tend to be much slower than resolving recursively on your own or using public DNS servers using anycast IPs (Quad9, Google, Cloudflare and co.).

> The macOS sandbox issue

Okay, sounds hard to solve as well. Thanks for your explanation, I understand, why my PR is not working in these, probably common, cases.

I thought about another solution to my issue, it may be harder to implement but maybe that's better in the long run: Currently, the exit node applies its "accept-dns=false" option to both itself & its exit's users. Would it be a solution to just apply this setting to itself and revert to using the settings in the online UI (if given) for its exit's users. Then if "Override local DNS" is off, the Canadian device would still use the European DNS (as the exit node just needs to send requests to it, which is possible on macOS/iOS as well). But if "Override local DNS" is given, the Canadian device would continue to use the settings declared in the Web UI. The exit node's OS can continue to ignore Tailscale's integrated 100.100.100.100 resolver and continue to resolve like when Tailscale is disabled (in case of the macOS/iOS app, the tailscale app just needs to differentiate between requests from local and from the exit node, which should be possible?). May this be a working solution?

**Comment by Zocker1999NET** [comment]
While changing my setup to work around my specific issue here until it is fixed, I found another problem which arises when it isn't possible to disable using the exit node's DNS configuration (which I proposed in my PR): If using a Pi-Hole & exit nodes together (using both together is proposed as a common use case [here](https://tailscale.com/kb/1114/pi-hole/)), you cannot differentiate between your devices as all DNS requests come from the same source IP (from the exit node's tailscale IP). This however disables one of Pi-Hole's main features: Allowing access to different domains from different clients. (I can open another issue for that if wanted. I did not found an issue already complaining about that.)

I personally can work around that as well (not that I want to), but I think this use case may be common enough for "normal" users so this should be changeable to users preference. And one solution to implement this may be my PR #5876 from above. This does not mean that this has to be the default & has to break working setups, that's why I would help testing & improving said PR until its good enough.

May this use case make my PR "more acceptable"? I accept another "rejection", however after thinking about the second use case, I think it is important enough to at least ask again and may also be important enough to be solved by the official Tailscale implementation.

**Comment by matrixuser0** [comment]
@Zocker1999NET I just saw your issue and PR, and funnily enough I had almost the same discussion with @DentonGentry a short while ago, although I proposed to disable ExitDNS on the exit node side to solve the same issue. Also see https://github.com/tailscale/tailscale/pull/5742

I think I'll take up @bradfitz's offer to discuss with them this use case where higher level of control over DNS resolution is required. Maybe the use case can be generalized sufficiently (taking into account also your use case) to condense it into a future product feature.

**Comment by dfnik** [comment]
@bradfitz @DentonGentry  as per our short discussion yesterday, would you generally entertain the idea of upstreaming a proposed feature where the tailscaled exit node optionally (envknob'd or otherwise configurable) attach EDNS Client Subnet information to forwarded queries, containing the full tailnet IPv4 / IPv6 address of the peer? This would provide a way for the upstream resolver to have all the information about the source of the query, in a standards compliant way.

ECS is a DNS extension well supported by other forwarders such as Google DNS, although for diffferent purposes (namely correctly geolocated upstream queries).

**Comment by matrixuser0** [comment]
@bradfitz @DentonGentry as per our short discussion yesterday, would you generally entertain the idea of upstreaming a proposed feature where the tailscaled exit node optionally (envknob'd or otherwise configurable) attach EDNS Client Subnet information to forwarded queries, containing the full tailnet IPv4 / IPv6 address of the peer? This would provide a way for the upstream resolver to have all the information about the source of the query, in a standards compliant way.

ECS is a DNS extension used also by other forwarders such as Google DNS, although for different purposes (namely correctly geolocated upstream queries) - for an example see https://developers.google.com/speed/public-dns/docs/ecs
